### PR TITLE
Fix doctests for Python 3 and PyPy

### PR DESCRIPTION
**Pull request by jamadden**
And enable them in tox and travis.

This is based off #27.

### CODE REVIEW

**Comment by jamadden** [comment]
(And rebased on master now that #27 is merged.)

File: docs/api/adapter.rst [hunk 1]
```diff
@@ -358,14 +358,20 @@ Let's register some adapters first:
    >>> gsm.registerAdapter(Comp, [None], I5, 'foo')
 
 Now we get all the adapters that are registered for ``ob`` that provide
-``I5``:
+``I5`` (note that on Python 2 the names will be ``unicode``):
```
Comment (tseaver) [inline docs/api/adapter.rst]: Can we use the term "text", and then note that text is unicode on Python2?

File: docs/socketexample.rst [hunk 2]
```diff
@@ -222,15 +223,26 @@ to know about all the adapters that convert a German to a US socket type:
...
+.. doctest::
+
+   >>> try:
+   ...    unicode = unicode
+   ... except NameError:
+   ...    unicode = str
+   >>> [isinstance(name, unicode) for name, _ in sockets]
+   [True, True, True]
+
+
```
Comment (tseaver) [inline docs/socketexample.rst]: Seems redundant given the example above, but if we included this, also use `text`.

File: docs/socketexample.rst [hunk 3]
```diff
@@ -552,7 +569,7 @@ need this method.
 .. doctest::
 
    >>> utils = list(zope.component.getAllUtilitiesRegisteredFor(IUSSocket))
-   >>> utils.sort()
+   >>> utils.sort(key=lambda x: type(x).__name__)
```
Comment (tseaver) [inline docs/socketexample.rst]: Prefer `sorted()` to creating a list and then sorting it in place (which I see you've done elsewhere).

File: docs/zcml.rst [hunk 4]
```diff
@@ -15,15 +15,15 @@ This helper will let us easily execute ZCML snippets:
...
 
-   >>> from cStringIO import StringIO
+   >>> from io import BytesIO
    >>> from zope.configuration.xmlconfig import xmlconfig
    >>> def runSnippet(snippet):
    ...     template = """\
    ...     <configure xmlns='http://namespaces.zope.org/zope'
    ...                i18n_domain="zope">
    ...     %s
    ...     </configure>"""
```
Comment (tseaver) [inline docs/zcml.rst]: `template` should be bytes, no?

File: docs/zcml.rst [hunk 5]
```diff
@@ -1129,6 +1131,10 @@ First we provide a stub configuration context:
 .. doctest::
 
    >>> import re, pprint
+   >>> try:
+   ...     from cStringIO import StringIO
+   ... except ImportError:
+   ...     from io import StringIO
```
Comment (tseaver) [inline docs/zcml.rst]: We should just use `io.StringIO` everywhere.

Comment (jamadden) [inline docs/zcml.rst]: You can't 😢 The `PrettyPrinter` calls `stream.write("(")` which is a str (aka bytes on Python 2).

File: docs/zcml.rst (hunk 4 above)
Comment (jamadden) [inline docs/zcml.rst]: No, native strings. They then get encoded into bytes later. (This was simpler than using `io.StringIO` and adding `u` prefixes everywhere for Python 2.)

File: docs/socketexample.rst (hunk 3 above)
Comment (jamadden) [inline docs/socketexample.rst]: The reason I didn't do that here was that the use of the key function means that `sorted` would either be one really long line or get broken into two, which looks bad (and is annoying hard to indent):
```pycon
>>> utils = list(zope.component.getAllUtilitiesRegisterFor(IUSSocket),
...               key=lambda x: type(x).__name__)
```

File: docs/zcml.rst (hunk 5 above)
Comment (tseaver) [inline docs/zcml.rst]: Like I needed more reasons to despise docstests.

File: docs/api/adapter.rst (hunk 1 above)
Comment (jamadden) [inline docs/api/adapter.rst]: OK, makes sense.

File: docs/socketexample.rst (hunk 3 above)
Comment (tseaver) [inline docs/socketexample.rst]: You could bind the result of `zope.component.getAllUtilitiesRegisterFor(IUSSocket)` to a temporary, e.g.:

```python
>>> all_utils = zope.component.getAllUtilitiesRegisteredFor(IUSSocket)
>>> utils = sorted(all_utils, key=lambda x: type(x).__name__)
```

Comment (jamadden) [inline docs/socketexample.rst]: That starts to obscure what's going on, IMO. I'll just use `sorted` and deal with the indentation.

**Pull request merged**